Acquire background activity for service worker termination

bnham opened the pull request:
#### b9c4c45e974285ed779ea8f4db66787b5e2f21d6
<pre>
Acquire background activity for service worker termination
<a href="https://bugs.webkit.org/show_bug.cgi?id=282319">https://bugs.webkit.org/show_bug.cgi?id=282319</a>
<a href="https://rdar.apple.com/138594750">rdar://138594750</a>

Reviewed by Youenn Fablet.

We are seeing some spurious ExceededCPULimit kills of WebContent processes associated with
background tabs. It looks like what is happening is that the process is suspended, and then for
whatever reason we have to terminate a service worker (e.g. in the logs we have, ITP is deleting
data for an origin and causing the associated service worker to be unregistered). NetworkProcess
then asks WebProcess to terminate the service worker directly, but the termination request never
even starts since the WebProcess is suspended.

To fix this, make NetworkProcess ask the UIProcess to start a background activity for WebProcess
that hosts the service worker before terminating it.

To test this, fix the existing SuspendAndTerminateWorker test to make sure that the process is fully
suspended before terminating the service worker. This needed to be fixed to take in to account the
process assertion cache.

* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::runningOrTerminatingCount const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::runningOrTerminatingServiceWorkerCountForTesting const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::terminateWorker):
(WebKit::WebSWServerToContextConnection::workerTerminated):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _runningOrTerminatingServiceWorkerCountForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::runningOrTerminatingServiceWorkerCountForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::workerTerminated):
* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate persistentDataStore]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, SuspendAndTerminateWorker)):

Canonical link: <a href="https://commits.webkit.org/286417@main">https://commits.webkit.org/286417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9846192cd5ba2259b901c16cb3aa5ba17010f43

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75906 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27172 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64077 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3229 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59514 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17681 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49397 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39874 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46797 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22676 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25499 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23013 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81867 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3275 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2072 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67742 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67050 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16712 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11000 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9126 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3225 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6031 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3246 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4184 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3253 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -2125,24 +2147,41 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
 
-    TestWebKitAPI::Util::runFor(1_s);
+    bool serviceWorkersAllTerminated = waitUntilEvaluatesToTrue([&]() {
+        __block bool done;
+        __block NSUInteger serviceWorkerCount;
+
+        [[WKWebsiteDataStore defaultDataStore] _runningOrTerminatingServiceWorkerCountForTesting:^(NSUInteger count) {
+            serviceWorkerCount = count;
+            done = true;
+        }];
+        TestWebKitAPI::Util::run(&done);
 
-    // Let's verify the WKWebView did not crash.
+        return !serviceWorkerCount;
+    }, 20);
+    EXPECT_TRUE(serviceWorkersAllTerminated);
```
It would probably be nice to replace this with something like:

```
    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return [processPool _serviceWorkerProcessCount] == 0; }));
```

But as far as I can tell, there is no connection between terminating a worker (even as a result of clearing the data store) and `_serviceWorkerProcessCount` decreasing. Is this expected?

youennf replied:
Yes, I think so. We remove the service worker process when there are no potential clients, see the call site of `removeContextConnectionIfPossible`

youennf commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1741,6 +1743,35 @@ void WebProcessProxy::requestTermination(ProcessTerminationReason reason)
+    WEBPROCESSPROXY_RELEASE_LOG(ProcessSuspension, "acquireServiceWorkerTerminationAssertion: Acquiring a background assertion to terminate service worker %llu", identifier.toUInt64());
+
+    m_serviceWorkersWithTerminationActivities.add(identifier);
+    m_dropServiceWorkerTerminationActivityTimer.startOneShot(WebCore::SWServer::defaultTerminationDelay + 1_s);
```
Do we really need this?
If the timer kicks in, it means stopping failed. In which case, the network process proxy should be notified of the unresponsiveness and `disableRemoteWorkers` be called.
How about just dropping the activity within `disableRemoteWorkers`?

bnham replied:
Will do

youennf commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1418,6 +1418,14 @@ - (void)_handleWebPushAction:(_WKWebPushAction *)webPushAction
+{
+    auto completionHandlerCopy = makeBlockPtr(completionHandler);
+    _websiteDataStore->runningOrTerminatingServiceWorkerCountForTesting([completionHandlerCopy = WTFMove(completionHandlerCopy)](auto result) {
+        completionHandlerCopy(result);
```
Maybe it would be useful for UIProcess to know whether a service worker process has any running service worker. That way, this process could exit in case of memory pressure. Or maybe the WebProcess should do this on its own, if it is not already happening.

bnham replied:
Yeah I was surprised there wasn't already a good way of finding this state without messaging NetworkProcess. But will look in to it in the future to not complicate this patch any further.

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2271,6 +2271,15 @@ void NetworkProcess::terminateRemoteWorkerContextConnectionWhenPossible(RemoteWo
+    if (!session)
+        return completionHandler(0);
+
+    completionHandler(session->ensureProtectedSWServer()->runningOrTerminatingCount());
```
I would tend to use ensureSWServer() or swServer(), hopefully this should be aligned with our C++ guidelines.

bnham replied:
Let me see what the Safer C++ checker says. I'm actually not sure why the first call in the chain is allowed since `session` is a raw pointer and `ensureProtectedSWServer()` (and `ensureSWServer()`) seem like non-trivial member functions.

youennf commented on Source/WebCore/workers/service/server/SWServer.h:
```diff
@@ -299,6 +299,8 @@ class SWServer : public RefCounted<SWServer>, public CanMakeWeakPtr<SWServer> {
 
     bool isProcessTerminationDelayEnabled() const { return m_isProcessTerminationDelayEnabled; }
 
+    unsigned runningOrTerminatingCount() { return m_runningOrTerminatingWorkers.size(); }
```
const?

youennf commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
```diff
@@ -244,9 +244,18 @@ void WebSWServerToContextConnection::fireBackgroundFetchClickEvent(ServiceWorker
 
 void WebSWServerToContextConnection::terminateWorker(ServiceWorkerIdentifier serviceWorkerIdentifier)
 {
+    protectedConnection()->protectedNetworkProcess()->protectedParentProcessConnection()->send(Messages::NetworkProcessProxy::AcquireServiceWorkerTerminationAssertion { webProcessIdentifier(), serviceWorkerIdentifier }, 0);
```
How about reusing NetworkProcessProxy::StartServiceWorkerBackgroundProcessing instead.

bnham replied:
I think I looked at that in the beginning but was concerned about overloading the meaning of `m_processingFunctionalEventCount`. But let me see if I can make that work, maybe I can rename the member variable and add some helper methods. I think I need to make sure I don't accidentally decrement that variable twice in case e.g. the termination event and termination timer both happen (in the case that the they happen around the same time and the term event gets stuck in a message queue while the timer fire is being processed.).

youennf commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
```diff
@@ -244,9 +244,18 @@ void WebSWServerToContextConnection::fireBackgroundFetchClickEvent(ServiceWorker
 {
+    protectedConnection()->protectedNetworkProcess()->protectedParentProcessConnection()->send(Messages::NetworkProcessProxy::AcquireServiceWorkerTerminationAssertion { webProcessIdentifier(), serviceWorkerIdentifier }, 0);
+
     send(Messages::WebSWContextManagerConnection::TerminateWorker(serviceWorkerIdentifier));
```
I would tend to use an async reply. `SWContextManager::terminateWorker` takes a callback (though it is a Function, worth checking).

youennf commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -592,6 +592,18 @@ void NetworkProcessProxy::terminateWebProcess(WebCore::ProcessIdentifier webProc
 
+void NetworkProcessProxy::acquireServiceWorkerTerminationAssertion(WebCore::ProcessIdentifier webProcessIdentifier, WebCore::ServiceWorkerIdentifier serviceWorkerIdentifier)
+{
+    if (auto process = WebProcessProxy::processForIdentifier(webProcessIdentifier))
```
We are apparently supposed to use `RefPtr`.

youennf commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -592,6 +592,18 @@ void NetworkProcessProxy::terminateWebProcess(WebCore::ProcessIdentifier webProc
+
+void NetworkProcessProxy::releaseServiceWorkerTerminationAssertion(WebCore::ProcessIdentifier webProcessIdentifier, WebCore::ServiceWorkerIdentifier serviceWorkerIdentifier)
+{
+    if (auto process = WebProcessProxy::processForIdentifier(webProcessIdentifier))
```
Ditto

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -2125,24 +2147,41 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+    // Due to the process assertion cache, the process can actually run for a little while after it
+    // enters the suspended state. Wait until the OS tells us the process is actually suspended.
+    pid_t pidBeforeTerminatingServiceWorker = [webView _webProcessIdentifier];
+    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return isPIDSuspended(pidBeforeTerminatingServiceWorker); }));
```
Ah nice, this is what I was missing, thanks for fixing this!

bnham commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -594,8 +594,10 @@ void NetworkProcessProxy::terminateWebProcess(WebCore::ProcessIdentifier webProc
 {
-    if (auto process = WebProcessProxy::processForIdentifier(processIdentifier))
+    if (auto process = WebProcessProxy::processForIdentifier(processIdentifier)) {
+        process->endServiceWorkerBackgroundProcessing();
```
I _think_ from reading the code this is necessary, although maybe I should fold in dropping the background processing activity in to `disableRemoteWorkers`?

bnham committed:
Acquire background activity for service worker termination
https://bugs.webkit.org/show_bug.cgi?id=282319
rdar://138594750

Reviewed by Youenn Fablet.

We are seeing some spurious ExceededCPULimit kills of WebContent processes associated with
background tabs. It looks like what is happening is that the process is suspended, and then for
whatever reason we have to terminate a service worker (e.g. in the logs we have, ITP is deleting
data for an origin and causing the associated service worker to be unregistered). NetworkProcess
then asks WebProcess to terminate the service worker directly, but the termination request never
even starts since the WebProcess is suspended.

To fix this, make NetworkProcess ask the UIProcess to start a background activity for WebProcess
that hosts the service worker before terminating it.

To test this, fix the existing SuspendAndTerminateWorker test to make sure that the process is fully
suspended before terminating the service worker. This needed to be fixed to take in to account the
process assertion cache.

* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::runningOrTerminatingCount const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::runningOrTerminatingServiceWorkerCountForTesting const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::terminateWorker):
(WebKit::WebSWServerToContextConnection::workerTerminated):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _runningOrTerminatingServiceWorkerCountForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::runningOrTerminatingServiceWorkerCountForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::workerTerminated):
* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate persistentDataStore]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, SuspendAndTerminateWorker)):

Canonical link: https://commits.webkit.org/286417@main
